src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,502 @@
+    s(Couple.of('s', Couple.of(" -+.0", argString.getValue()))),
+    t(Couple.of('t', Couple.of("-", argBool.getValue()))),
+    T(Couple.of('T', Couple.of("-", anyType.getValue()))),
+    U(Couple.of('U', Couple.of("-#", argRune.getValue() | argInt.getValue()))),
+    V(Couple.of('v', Couple.of(" -+.0#", anyType.getValue()))),
+    x(Couple.of('x', Couple.of(" -+.0#", argRune.getValue() | argInt.getValue() | argString.getValue()))),
+    X(Couple.of('X', Couple.of(" -+.0#", argRune.getValue() | argInt.getValue() | argString.getValue())));
+
+    private Couple myVal;
+
+    PrintVerb(Couple val) {
+      myVal = val;
+    }
+
+    public Couple getValue() {
+      return myVal;
+    }
+
+    public char getVerb() {
+      return (char)myVal.getFirst();
+    }
+
+    @NotNull
+    public String getFlags() {
+      return (String)((Couple)myVal.getSecond()).getFirst();
+    }
+
+    public int getMask() {
+      return (int)((Couple)myVal.getSecond()).getSecond();
+    }
+
+    @Nullable
+    public static PrintVerb getByVerb(char verb) {
+      for (PrintVerb v : values()) {
+        if (verb == v.getVerb()) return v;
+      }
+      return null;
+    }
+  }
+
+  private ProblemsHolder myProblemsHolder;
+  private GoCallExpr myCallExpr;
+  private GoFunctionOrMethodDeclaration myDeclaration;
+
+  public GoPlaceholderChecker(ProblemsHolder problemsHolder, GoCallExpr callExpr, GoFunctionOrMethodDeclaration declaration) {
+    myProblemsHolder = problemsHolder;
+    myCallExpr = callExpr;
+    myDeclaration = declaration;
+  }
+
+  public void checkPrint() {
+    // TODO florin: Implement the print function checking
+    // Ref code: https://github.com/golang/go/blob/79f7ccf2c3931745aeb97c5c985b6ac7b44befb4/src/cmd/vet/print.go#L587
+  }
+
+  public void checkPrintf() {
+    int placeholderPosition = getPlaceholderPosition(myDeclaration);
+    List<GoExpression> arguments = myCallExpr.getArgumentList().getExpressionList();
+    if (placeholderPosition < 0 || arguments.size() <= placeholderPosition) return;
+
+    GoExpression placeholder = arguments.get(placeholderPosition);
+    if (!GoTypeUtil.isString(placeholder.getGoType(null))) {
+      String message = "Value used for formatting text does not appear to be a string";
+      myProblemsHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
+      return;
+    }
+
+    String placeholderText = resolve(placeholder);
+    if (placeholderText == null) return;
+
+    // We always receive the text with double quotes at the beginning and at the end so remove them
+    placeholderText = placeholderText.substring(1, placeholderText.length() - 1);
+
+    if (!hasPlaceholder(placeholderText) && arguments.size() > placeholderPosition) {
+      String message = "Value used for formatting text does not appear to contain a placeholder";
+      myProblemsHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
```

Comment:
Still strange dependency on problems holder here. You have a parser which just parses some string.
First of all it's weird that the parser knows about problemsHolder. In the second, it means that you can parse printf placeholder only in inspections, so no completion support in future. In the third, what is the real responsibility of GoPlaceholderChecker? If it's a Checker it should check and return the result (success or not). If it's parser it should parse and return parsed representation of printf argument (e.g. a Map). If it's inspection, it should visit elements and create problem descriptors for some of them. You're trying to create a class that knows everything and does all these three things. It's hard to read, hard to modify and hard to reuse. Try to keep single responsibility for each class and method.